aspects/schema_test.go

```diff
@@ -0,0 +1,355 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package aspects_test
+
+import (
+	"github.com/snapcore/snapd/aspects"
+	. "gopkg.in/check.v1"
+)
+
+type schemaSuite struct{}
+
+var _ = Suite(&schemaSuite{})
+
+func (*schemaSuite) TestTopLevelFailsWithoutSchema(c *C) {
+	schemaStr := []byte(`{
+  "keys": "string"
+}`)
+
+	_, err := aspects.ParseSchema(schemaStr)
+	c.Assert(err, ErrorMatches, `cannot parse top level schema: must have a "schema" constraint`)
+}
+
+func (*schemaSuite) TestSchemaMustBeMap(c *C) {
+	schemaStr := []byte(`["foo"]`)
+
+	_, err := aspects.ParseSchema(schemaStr)
+	c.Assert(err, ErrorMatches, `cannot parse top level schema: must be a map`)
+}
+
+func (*schemaSuite) TestMapWithSchemaConstraint(c *C) {
+	schemaStr := []byte(`{
+  "schema": {
+    "snaps": {
+			"type": "map",
```

Comment:
My bad, some were indented with 2 spaces and others with a tab. I missed it because I use 2 widths for tabs so it looks the same in my editor but evidently not in Github. I moved all the JSON snippets to tabs so it should look consistent now